File: library/src/main/java/com/mikepenz/materialdrawer/holder/BadgeStyle.java
package com.mikepenz.materialdrawer.holder;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.widget.TextView;

import com.mikepenz.materialdrawer.R;
import com.mikepenz.materialdrawer.model.utils.BadgeDrawableBuilder;

import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.annotation.DimenRes;
import androidx.annotation.Dimension;
import androidx.annotation.DrawableRes;
import androidx.core.view.ViewCompat;

import static androidx.annotation.Dimension.DP;
import static androidx.annotation.Dimension.PX;

/**
 * Class to allow defining a BadgeStyle for the `BadgeDrawerItem`
 */
public class BadgeStyle {
    private int mGradientDrawable = R.drawable.material_drawer_badge;
    private Drawable mBadgeBackground;
    private ColorHolder mColor;
    private ColorHolder mColorPressed;
    private ColorHolder mTextColor;
    private ColorStateList mTextColorStateList;
    private DimenHolder mCorners;
    private DimenHolder mPaddingTopBottom = DimenHolder.fromDp(2); //2 looks best
    private DimenHolder mPaddingLeftRight = DimenHolder.fromDp(3); //3 looks best
    private DimenHolder mMinWidth = DimenHolder.fromDp(20); //20 looks nice

    public int getGradientDrawable() {
        return mGradientDrawable;
    }

    public BadgeStyle withGradientDrawable(@DrawableRes int gradientDrawable) {
        this.mGradientDrawable = gradientDrawable;
        this.mBadgeBackground = null;
        return this;
    }

    public Drawable getBadgeBackground() {
        return mBadgeBackground;
    }

    public BadgeStyle withBadgeBackground(Drawable badgeBackground) {
        this.mBadgeBackground = badgeBackground;
        this.mGradientDrawable = -1;
        return this;
    }

    public ColorHolder getColor() {
        return mColor;
    }

    public BadgeStyle withColor(@ColorInt int color) {
        this.mColor = ColorHolder.fromColor(color);
        return this;
    }

    public BadgeStyle withColorRes(@ColorRes int color) {
        this.mColor = ColorHolder.fromColorRes(color);
        return this;
    }

    public ColorHolder getColorPressed() {
        return mColorPressed;
    }

    public BadgeStyle withColorPressed(@ColorInt int colorPressed) {
        this.mColorPressed = ColorHolder.fromColor(colorPressed);
        return this;
    }

    public BadgeStyle withColorPressedRes(@ColorRes int colorPressed) {
        this.mColorPressed = ColorHolder.fromColorRes(colorPressed);
        return this;
    }

    public ColorHolder getTextColor() {
        return mTextColor;
    }

    public BadgeStyle withTextColor(@ColorInt int textColor) {
        this.mTextColor = ColorHolder.fromColor(textColor);
        return this;
    }

    public BadgeStyle withTextColorRes(@ColorRes int textColor) {
        this.mTextColor = ColorHolder.fromColorRes(textColor);
        return this;
    }

    public BadgeStyle withTextColorStateList(ColorStateList textColorStateList) {
        this.mTextColor = null;
        this.mTextColorStateList = textColorStateList;
        return this;
    }

    public DimenHolder getCorners() {
        return mCorners;
    }

    public BadgeStyle withCorners(@Dimension(unit = PX) int cornersPx) {
        this.mCorners = DimenHolder.fromPixel(cornersPx);
        return this;
    }

    public BadgeStyle withCornersDp(@Dimension(unit = DP) int corners) {
        this.mCorners = DimenHolder.fromDp(corners);
        return this;
    }

    public BadgeStyle withCorners(DimenHolder corners) {
        this.mCorners = corners;
        return this;
    }

    public DimenHolder getPaddingLeftRight() {
        return mPaddingLeftRight;
    }

    public BadgeStyle withPaddingLeftRightPx(@Dimension(unit = PX) int paddingLeftRight) {
        this.mPaddingLeftRight = DimenHolder.fromPixel(paddingLeftRight);
        return this;
    }

    public BadgeStyle withPaddingLeftRightDp(@Dimension(unit = DP) int paddingLeftRight) {
        this.mPaddingLeftRight = DimenHolder.fromDp(paddingLeftRight);
        return this;
    }

    public BadgeStyle withPaddingLeftRightRes(@DimenRes int paddingLeftRight) {
        this.mPaddingLeftRight = DimenHolder.fromResource(paddingLeftRight);
        return this;
    }

    public DimenHolder getPaddingTopBottom() {
        return mPaddingTopBottom;
    }

    public BadgeStyle withPaddingTopBottomPx(@Dimension(unit = PX) int paddingTopBottom) {
        this.mPaddingTopBottom = DimenHolder.fromPixel(paddingTopBottom);
        return this;
    }

    public BadgeStyle withPaddingTopBottomDp(@Dimension(unit = DP) int paddingTopBottom) {
        this.mPaddingTopBottom = DimenHolder.fromDp(paddingTopBottom);
        return this;
    }

    public BadgeStyle withPaddingTopBottomRes(@DimenRes int paddingTopBottom) {
        this.mPaddingTopBottom = DimenHolder.fromResource(paddingTopBottom);
        return this;
    }

    public BadgeStyle withPadding(@Dimension(unit = PX) int padding) {
        this.mPaddingLeftRight = DimenHolder.fromPixel(padding);
        this.mPaddingTopBottom = DimenHolder.fromPixel(padding);
        return this;
    }

    public BadgeStyle withPadding(DimenHolder padding) {
        this.mPaddingLeftRight = padding;
        this.mPaddingTopBottom = padding;
        return this;
    }

    public DimenHolder getMinWidth() {
        return mMinWidth;
    }

    public BadgeStyle withMinWidth(@Dimension(unit = PX) int minWidth) {
        this.mMinWidth = DimenHolder.fromPixel(minWidth);
        return this;
    }

    public BadgeStyle withMinWidth(DimenHolder minWidth) {
        this.mMinWidth = minWidth;
        return this;
    }

    public BadgeStyle() {
    }

    public BadgeStyle(@ColorInt int color, @ColorInt int colorPressed) {
        this.mColor = ColorHolder.fromColor(color);
        this.mColorPressed = ColorHolder.fromColor(colorPressed);
    }

    public BadgeStyle(@DrawableRes int gradientDrawable, @ColorInt int color, @ColorInt int colorPressed, @ColorInt int textColor) {
        this.mGradientDrawable = gradientDrawable;
        this.mColor = ColorHolder.fromColor(color);
        this.mColorPressed = ColorHolder.fromColor(colorPressed);
        this.mTextColor = ColorHolder.fromColor(textColor);
    }

    public void style(TextView badgeTextView) {
        style(badgeTextView, null);
    }

    public void style(TextView badgeTextView, ColorStateList colorStateList) {
        Context ctx = badgeTextView.getContext();
        //set background for badge
        if (mBadgeBackground == null) {
            ViewCompat.setBackground(badgeTextView, new BadgeDrawableBuilder(this).build(ctx));
        } else {
            ViewCompat.setBackground(badgeTextView, mBadgeBackground);
        }

        //set the badge text color
        if (mTextColor != null) {
            ColorHolder.applyToOr(mTextColor, badgeTextView, null);
        } else if (mTextColorStateList != null) {
            badgeTextView.setTextColor(mTextColorStateList);
        } else if (colorStateList != null) {
            badgeTextView.setTextColor(colorStateList);
        }

        //set the padding
        int paddingLeftRight = mPaddingLeftRight.asPixel(ctx);
        int paddingTopBottom = mPaddingTopBottom.asPixel(ctx);
        badgeTextView.setPadding(paddingLeftRight, paddingTopBottom, paddingLeftRight, paddingTopBottom);

        //set the min width
        badgeTextView.setMinWidth(mMinWidth.asPixel(ctx));
    }
}

File: library/src/main/java/com/mikepenz/materialdrawer/holder/BadgeStyle.kt
package com.mikepenz.materialdrawer.holder

import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.core.view.ViewCompat
import com.mikepenz.materialdrawer.R
import com.mikepenz.materialdrawer.model.utils.BadgeDrawableBuilder

/**
 * Class to allow defining a BadgeStyle for the `BadgeDrawerItem`
 */
open class BadgeStyle {
    /** defines the drawable to use to define the rounded corners */
    var gradientDrawable = R.drawable.material_drawer_badge

    /** defines the background drawable */
    var badgeBackground: Drawable? = null

    /** the default color */
    var color: ColorHolder? = null

    /** the pressed color */
    var colorPressed: ColorHolder? = null

    /**
     * the text size
     * NOTE: Will only apply on 21+, also if applying, ensure to apply to all views
     */
    var textSizeSp: Float? = null
    private var _textColor: ColorHolder? = null

    /** defines the default text color */
    var textColor: ColorHolder?
        get() = _textColor
        set(value) {
            _textColorStateList = null
            _textColor = value
        }

    private var _textColorStateList: ColorStateList? = null

    /** defines the alternative text color state list */
    var textColorStateList: ColorStateList?
        get() = _textColorStateList
        set(value) {
            _textColor = null
            _textColorStateList = value
        }

    /** the corner radious */
    var corners: DimenHolder? = null

    /** dcustom padding to the bottom (default 2dp) */
    var paddingTopBottom = DimenHolder.fromDp(2)

    /** custom padding to the right (default 3dp) */
    var paddingLeftRight = DimenHolder.fromDp(3)

    /** the min width to set (default 20dp) */
    var minWidth = DimenHolder.fromDp(20)

    /**
     * elevation to apply on the view
     * NOTE: Will only apply on 21+, also if applying, ensure to apply to all views
     */
    var elevation: DimenHolder? = null

    constructor()

    constructor(@ColorInt color: Int, @ColorInt colorPressed: Int) {
        this.color = ColorHolder.fromColor(color)
        this.colorPressed = ColorHolder.fromColor(colorPressed)
    }

    constructor(@DrawableRes gradientDrawable: Int, @ColorInt color: Int, @ColorInt colorPressed: Int, @ColorInt textColor: Int) {
        this.gradientDrawable = gradientDrawable
        this.color = ColorHolder.fromColor(color)
        this.colorPressed = ColorHolder.fromColor(colorPressed)
        this.textColor = ColorHolder.fromColor(textColor)
    }

    /** styles theprovided textView with this style, and the provided colorStateList */
    @JvmOverloads
    open fun style(badgeTextView: TextView, colorStateList: ColorStateList? = null) {
        val ctx = badgeTextView.context
        //set background for badge
        if (badgeBackground == null) {
            ViewCompat.setBackground(badgeTextView, BadgeDrawableBuilder(this).build(ctx))
        } else {
            ViewCompat.setBackground(badgeTextView, badgeBackground)
        }

        val textSizeSp = textSizeSp
        if (textSizeSp != null) {
            badgeTextView.textSize = textSizeSp
        } else {
            // keep the size it is defined in the layout
        }

        //set the badge text color
        when {
            textColor != null -> textColor?.applyToOr(badgeTextView, null)
            textColorStateList != null -> badgeTextView.setTextColor(textColorStateList)
            colorStateList != null -> badgeTextView.setTextColor(colorStateList)
        }

        //set the padding
        val paddingLeftRight = this.paddingLeftRight.asPixel(ctx)
        val paddingTopBottom = this.paddingTopBottom.asPixel(ctx)
        badgeTextView.setPadding(paddingLeftRight, paddingTopBottom, paddingLeftRight, paddingTopBottom)

        //set the min width
        badgeTextView.minWidth = minWidth.asPixel(ctx)

        // set elevation if expected
        elevation?.let { elev ->
            ViewCompat.setElevation(badgeTextView, elev.asPixel(ctx).toFloat())
        }
    }
}
